src/memory.rs

```diff
@@ -133,69 +154,79 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
         }
     }
 
-    pub fn allocate(&mut self, size: usize) -> EvalResult<'tcx, Pointer> {
+    pub fn allocate(&mut self, size: usize, align: usize) -> EvalResult<'tcx, Pointer> {
         if size == 0 {
             return Ok(Pointer::zst_ptr());
         }
+        // make sure we can offset the result pointer by the worst possible alignment
+        // this allows cheaply checking for alignment directly in the pointer
+        let least_aligned_size = size + align;
         if self.memory_size - self.memory_usage < size {
             return Err(EvalError::OutOfMemory {
-                allocation_size: size,
+                allocation_size: least_aligned_size,
                 memory_size: self.memory_size,
                 memory_usage: self.memory_usage,
             });
         }
         self.memory_usage += size;
         let alloc = Allocation {
-            bytes: vec![0; size],
+            bytes: vec![0; least_aligned_size],
             relocations: BTreeMap::new(),
-            undef_mask: UndefMask::new(size),
+            undef_mask: UndefMask::new(least_aligned_size),
+            align: align,
         };
         let id = self.next_id;
         self.next_id.0 += 1;
         self.alloc_map.insert(id, alloc);
         Ok(Pointer {
             alloc_id: id,
-            offset: 0,
+            // offset by the alignment, so larger accesses will fail
```

Comment:
Maybe we should just cache the allocation's alignment in the pointer? Not sure...